src/main/java/nextstep/subway/domain/Sections.java

```diff
@@ -0,0 +1,110 @@
+        if (Objects.nonNull(sectionByDownStation)) {
+            addBetweenSection(sectionByDownStation, sectionByDownStation.getUpStation(), upStation, downStation, sectionByDownStation.getDistance() - distance, distance);
+        }
+    }
+
+    private void addBetweenSection(Section section, Station upStation, Station middleStation, Station downStation, int distanceBetweenUpAndMiddleStation, int distanceBetweenMiddleAndDownStation) {
+        if (distanceBetweenUpAndMiddleStation <= 0 || distanceBetweenMiddleAndDownStation <= 0) {
+            throw new IllegalArgumentException();
+        }
+
+        this.sections.remove(section);
+        this.sections.add(new Section(section.getLine(), upStation, middleStation, distanceBetweenUpAndMiddleStation));
+        this.sections.add(new Section(section.getLine(), middleStation, downStation, distanceBetweenMiddleAndDownStation));
+    }
+
+    public void deleteSection(final Station station) {
+        if (this.sections.size() == 0) {
+            throw new IllegalArgumentException();
+        }
+
+        if (!isDownStationEndpoint(station)) {
+            throw new IllegalArgumentException();
+        }
+
+        this.sections.remove(getSectionByDownStation(station));
+    }
+
+    public boolean isDownStationEndpoint(Station station) {
+        return getDownStationEndpoint().equals(station);
+    }
+
+    private boolean isUpStationEndpoint(Station station) {
+        return geUpStationEndpoint().equals(station);
+    }
+
+    public Station geUpStationEndpoint() {
+        List<Station> upStations = sections.stream().map(Section::getUpStation).collect(Collectors.toList());
+        List<Station> downStations = sections.stream().map(Section::getDownStation).collect(Collectors.toList());
+
+        return upStations.stream().filter(it -> !downStations.contains(it)).findAny().orElse(null);
+    }
+
+    private Station getDownStationEndpoint() {
+        List<Station> upStations = sections.stream().map(Section::getUpStation).collect(Collectors.toList());
+        List<Station> downStations = sections.stream().map(Section::getDownStation).collect(Collectors.toList());
+
+        return downStations.stream().filter(it -> !upStations.contains(it)).findAny().orElseThrow(IllegalArgumentException::new);
+    }
+
+    public Section getSectionByUpStation(Station station) {
+        return this.sections.stream().filter(it -> it.getUpStation().equals(station)).findAny().orElse(null);
+    }
+
+    private Section getSectionByDownStation(Station station) {
+        return this.sections.stream().filter(it -> it.getDownStation().equals(station)).findAny().orElse(null);
+    }
+
+    private List<Station> allStations() {
+        List<Station> allStations = new ArrayList<>();
+        List<Station> upStations = sections.stream().map(Section::getUpStation).collect(Collectors.toList());
+        Station haHang = getDownStationEndpoint();
+
+        allStations.addAll(upStations);
```

Comment:
```suggestion
List<Station> allStations = new ArrayList<>(upStations);
```

이런 형태로 구현할 수 있을 것 같아요

Reply:
인텔리제이에서도 해당 방식으로 수정을 권유해주고 있어서 기존 방식과 제안해주신 방식 모두 고민해보았었는데, 빈 객체를 만들고 추가하는 방식이 가독성이 더 좋지 않을까 라는 생각을 했었던 것 같아요! 

수정해두었습니다!